kratos/processes/calculate_nodal_distance_to_skin_process.cpp

```diff
@@ -92,7 +101,35 @@ void CalculateNodalDistanceToSkinProcess::Execute()
         });
     };
 
-    const auto* p_distance_lambda = mHistoricalVariable ? &distance_lambda_historical : &distance_lambda_non_historical;
+    const std::function<void(ModelPart::NodesContainerType& rNodes, GeometricalObjectsBins& rBins)> distance_lambda_historical_save_skin = [this](ModelPart::NodesContainerType& rNodes, GeometricalObjectsBins& rBins) {
+        const auto& r_variable = *mpDistanceVariable;
+        for (Node& rNode: rNodes) {
+            auto result = rBins.SearchNearest(rNode);
+            const double value = result.GetDistance();
+            auto p_result = result.Get().get();
+            if (p_result->GetValue(r_variable) < value) {
+                p_result->SetValue(r_variable, value);
+                p_result->Set(VISITED);
+            }
+            rNode.FastGetSolutionStepValue(r_variable) = value;
+        }
+    };
+    const std::function<void(ModelPart::NodesContainerType& rNodes, GeometricalObjectsBins& rBins)> distance_lambda_non_historica_save_skin = [this](ModelPart::NodesContainerType& rNodes, GeometricalObjectsBins& rBins) {
+        const auto& r_variable = *mpDistanceVariable;
+        for (Node& rNode: rNodes) {
+            auto result = rBins.SearchNearest(rNode);
+            const double value = result.GetDistance();
+            auto p_result = result.Get().get();
+            if (p_result->GetValue(r_variable) < value) {
+                p_result->SetValue(r_variable, value);
+                p_result->Set(VISITED);
+            }
+            rNode.SetValue(r_variable, value);
+        }
+    };
```

Comment:
Wouldn't be cleaner and with less duplication if we only define a template method SetNodalValue and specialize it with historical and non historical template arguments

Reply:
Done